secrets/store.go

```diff
@@ -59,62 +58,63 @@ func (s *Store) parser(r io.Reader) (interface{}, error) {
 		return nil, err
 	}
 
-	if s.secretHandlerFunc != nil {
-		s.secretHandlerFunc(secrets)
-	}
+	s.secretHandlerFunc(secrets)
 
 	return secrets, nil
 }
 
-// SecretHandler creates the middleware chain.
+// secretHandler creates the middleware chain.
 func (s *Store) secretHandler(middlewares ...SecretMiddleware) {
-	if s.secretHandlerFunc == nil {
-		s.secretHandlerFunc = nopSecretHandlerFunc
-	}
-
 	for _, m := range middlewares {
 		s.secretHandlerFunc = m(s.secretHandlerFunc)
 	}
 }
 
-// GetSimpleSecret loads secrets from watcher, and fetches a simple secret from secrets
-func (s *Store) GetSimpleSecret(path string) (SimpleSecret, error) {
-	secrets, ok := s.watcher.Get().(*Secrets)
-	if !ok {
-		return SimpleSecret{}, fmt.Errorf("unexpected type %T", secrets)
-	}
+func (s *Store) getSecrets() *Secrets {
+	return s.watcher.Get().(*Secrets)
+}
 
-	return secrets.GetSimpleSecret(path)
+// Stop stops the underlying filewatcher and release associated resources.
+//
+// After Stop is called, you won't get any updates to the secret file,
+// but can still access the secrets as they were before Stop is called.
+//
+// It's OK to call Stop multiple times. Calls after the first one are no-ops.
+func (s *Store) Stop() {
+	s.watcher.Stop()
 }
 
-// GetVersionedSecret loads secrets from watcher, and fetches a versioned secret from secrets
-func (s *Store) GetVersionedSecret(path string) (VersionedSecret, error) {
-	secrets, ok := s.watcher.Get().(*Secrets)
-	if !ok {
-		return VersionedSecret{}, fmt.Errorf("unexpected type %T", secrets)
-	}
+// AddMiddlewares registers new middlewares to the store.
+//
+// Every AddMiddlewares call will cause all already registered middlewares to be
+// called again with the latest data.
+//
+// AddMiddlewares call is not thread-safe, it should not be called concurrently.
+func (s *Store) AddMiddlewares(middlewares ...SecretMiddleware) {
+	s.secretHandler(middlewares...)
+	s.secretHandlerFunc(s.getSecrets())
+}
 
-	return secrets.GetVersionedSecret(path)
+// GetSimpleSecret loads secrets from watcher, and fetches a simple secret from secrets
+func (s Store) GetSimpleSecret(path string) (SimpleSecret, error) {
+	return s.getSecrets().GetSimpleSecret(path)
 }
 
-// GetCredentialSecret loads secrets from watcher, and fetches a credential secret from secrets
-func (s *Store) GetCredentialSecret(path string) (CredentialSecret, error) {
-	secrets, ok := s.watcher.Get().(*Secrets)
```

Comment:
I don't think we should remove the type check. I don't like making assumptions based on upstream code that may or may not change in the future. If the type would be set in stone just a few lines above I'd be okay with it but this is a different package.

Reply:
1. This is the behavior of stdlib's `atomic.Value`, I'm pretty sure that's part of the the go 1 backward compatibility guarantee
2. We control the parser implementation, and there's no reason we would return a different type in the parser.

Reply:
But on these lines I don't see that it's using `atomic.Value`, it's referencing a different package. It's just defensive programming, doesn't really hurt to keep the practice of checking the type when you type assert. The opposite, does.

Reply:
Even though, the second point still stands, that we control what are the values stored in the same package (parser implementation).

Reply:
See also: https://pkg.go.dev/github.com/reddit/baseplate.go/filewatcher?tab=doc#Parser & https://pkg.go.dev/github.com/reddit/baseplate.go/filewatcher?tab=doc#Result.Get

Reply:
Okay, sounds good!